gravitational/teleport
# Fixes dissonance between `disconnect_expired_cert` vs `require_session_mfa`

Adds ssh/tls cert extensions
```
// CertExtensionMFAVerifiedSessionExpires is the RFC3339 timestamp representing the hard
// deadline of the session on a certificates issued after an MFA check.
// See https://github.com/gravitational/teleport/issues/18544.
CertExtensionMFAVerifiedSessionExpires = "mfa-verified-session-expires"

// MFAVerifiedSessionExpiresASN1ExtensionOID is the RFC3339 timestamp representing the hard
// deadline of the session on a certificates issued after an MFA check.
// See https://github.com/gravitational/teleport/issues/18544.
MFAVerifiedSessionExpiresASN1ExtensionOID = asn1.ObjectIdentifier{1, 3, 9999, 2, 12}
```

These get set when short lived (<1m) single-use MFA certs are issued, and ultimately passed to the connection monitor as the `DisconnectExpiredCert` value, which fixes https://github.com/gravitational/teleport/issues/18544

### Manual testing

See instructions in [webapps README](https://github.com/gravitational/webapps/blob/1e133f2e4975dc91db6f2cf1c8a14dcd7ecdf770/README.md?plain=1#L253-L271) for local mfa development. My auth service looks like
```yaml
auth_service:
  enabled: "yes"
  listen_addr: 0.0.0.0:3025
  authentication:
    second_factor: optional #off
    require_session_mfa: yes
    webauthn:
      rp_id: proxy.127.0.0.1.nip.io
  disconnect_expired_cert: yes
```

To confirm the force-disconnection time on mfa sessions, I added the line

```go
fmt.Printf("Monitor force-disconnecting in %v\n", discTime)
```

below this one in `monitor.go`
https://github.com/gravitational/teleport/blob/b571efe990b1b2db089199f002f4d83075eee335/lib/srv/monitor.go#L181

Before these changes it would show ~56seconds, whereas now it shows a time closer to the cluster's default ttl (12h in my case).

**Behavior manually confirmed for**
- [x] SSH (@ibeckermayer)
- [x] Desktop Access (@ibeckermayer)
- [x] Kubernetes (@tigrato)
- [x] App Access (@mdwn): will be double checked once https://github.com/gravitational/teleport/pull/18607#issuecomment-1323967391 is addressed, we can proceed with merging this fix for now.
- [x] DB Access (@smallinsky)
